### DISCUSSION TITLE
Converting between DynamicVector[UI8], DynamicVector[Int], and Buffer[x, UI8] in a zero copy way

### DISCUSSION THREAD

**Discussion by sa-** [modular/modular#127] [Q&A] [answered]
Hey there, 

I'm trying to convert the type of some memory between DynamicVector[UI8], DynamicVector[Int], and Buffer[x, UI8] and I would like to do this in a zero copy way. I'm having a hard time making it work and I would appreciate it if someone has an example on how to do this.

Thanks in advance!

**Comment by czheo** [comment]
Does this work?
``` py
from Vector import DynamicVector

var int_vec = DynamicVector[Int]()
int_vec.push_back(123)
print(int_vec[0]) # 123

var ui8_vec = DynamicVector[UI8](int_vec.data.bitcast[UI8](), int_vec.size)
print(ui8_vec[0]) #123
```

**Reply by sa-** [reply]
Is `DynamicVector.size` in bytes?

**Reply by sa-** [reply]
Or is it the number of elements with the associated type? The types could have different bitwidths

**Reply by sa-** [reply]
Looks like it's the latter because `int_vec.size` is 1, and if the number you push into the `int_vec` is greater than 255 it's going to be problematic

**Reply by czheo** [reply]
> if the number you push into the int_vec is greater than 255 it's going to be problematic

That's because >255 is out of range of `UI8`.

**Reply by sa-** [reply]
Okay maybe that was a bad example, but it's problematic for more than that reason

**Reply by sa-** [reply]
`ui8.size` in the snippet that you posted is `1`, but it should be `8`

**Comment by sa-** [comment, marked as answer]
I've figured this out. Thanks to @czheo for helping me figure this out
```python
from TargetInfo import sizeof

fn reinterpret_dynamic_vector[Tsrc: AnyType, Tdest: AnyType](
        src: DynamicVector[Tsrc]
) -> DynamicVector[Tdest]:
    let tsrc_bitwidth = sizeof[Tsrc]()
    let tdest_bitwidth = sizeof[Tdest]()
    
    let num_bytes_src = tsrc_bitwidth * src.size
    if num_bytes_src % tdest_bitwidth != 0:
        print("Cannot divide up the bits for the new type")
    let dest_num_elements = num_bytes_src // tdest_bitwidth
    return DynamicVector[Tdest](src.data.bitcast[Tdest](), dest_num_elements)
    
var int_vec = DynamicVector[Int]()
int_vec.push_back(-1)
print(int_vec[0]) # output: -1

var ui8_vec = reinterpret_dynamic_vector[Int, UI8](int_vec)

print(ui8_vec.size) # output: 8
for i in range(ui8_vec.size):
    print(ui8_vec[i]) # outout: 8x 255
```
